doc/developer/design/20210811_catalog_only_mode.md

```diff
@@ -0,0 +1,115 @@
+into tables; these are the potentially prohibited actions. To support this enforcement,
+we'll add a new field to the `Coordinator`:
+
+```rust
+pub struct Coordinator {
+    ...
+    /// When `None`, all user objects can create/access dataflows; when `Some`, only the objects
+    /// with an ID in the hashmap are allowed.
+    catalog_only_mode_allow: Option<HashSet<GlobalId>>,
+}
+```
+
+When booting normally, we'll set this field to `None`, and
+`Some(HashSet::new())` when booting in catalog-only mode.
+
+Then, whenever a user or the system attempts to perform a potentially prohibited
+action on an object, we'll allow the action if any of the conditions are true:
+- `catalog_only_mode_allow.is_none()`
+- The `GlobalId` is not a user-created object
+- Reads/peeks/`SELECT`s occur on the fast path, i.e. they are coming from a
+  materialization, and a selected index is enabled.
+  
+  Note that:
+  - This might mean that the optimizer chooses a set of indexes for a query that
+    have not been enabled, even though others have been, so we will message which
+    indexes need to be enabled or dropped for the query to potentially work.
+    Emphasis on might; I'm not sure how optimistic or pessimistic the generated
+    sets of indexes are.
+  - Querying unmaterialized sources with unmaterialized views becomes
+    impossible. However, users could create a new materialized view, and then
+    enable it, to query unmaterialized sources in catalog-only mode.
+- Inserts occur on tables IDs that have been implicitly enabled by having a
+  covering index enabled.
+
+### Enabling user arrangements
+
+When we receive a `ALTER INDEX ... SET ENABLED` command from the user, we'll:
+
+1. Ensure the named object is an index and error if not.
+1. Check if `catalog_only_mode_allow.is_none()` and return OK if so.
+1. Get the index's `GlobalId`.
+1. Check if the unwrapped `catalog_only_mode_allow` contains the index's ID,
+   indicating it's already been restarted and return OK if so.
+1. Use something akin following code to ship the dataflow we previously
+   prohibited:
+    ```rust
+    let df = self.dataflow_builder().build_index_dataflow(id);
+    self.ship_dataflow(df);
+    ```
+1. Add the index's `GlobalId` to `catalog_only_mode_allow`.
+1. If the named index is an index on a table, add the table's `GlobalId` to
+   `catalog_only_mode_allow`. This is necessary because sequencing volatile
+   inserts (i.e. writing to tables) doesn't declare which index/arrangement will
+   get used, so we have to save this inference. Also, all indexes in Materialize
+   are covering, so this should continue to allow writes to all columns.
+   
+   If my memory is mistaken and this isn't the case (i.e. indexes can offer
+   partial cover), we'll only write the table's `GlobalId` to
+   `catalog_only_mode_allow` when a covering index gets enabled.
+
+## Alternatives
+
+No alternative designs were seriously considered.
```

Comment:
What about providing a feature to have MZ rollback or disable the last _N_ DDLs? e.g. `materialized --disable-recent-ddls 1` would make Materialize disable last DDL in the catalog. It seems likely that one of the more recent DDLs (probably the most recent or second most recent) instigated the OOM loop.

Reply:
That's an interesting take! I poked around the code a for a while and I believe we'd still want to load everything in the catalog, though, so we could still drop the resources; and at that point, what we'd need to do is stop the dataflows from spinning up, so seems like it reduces to this proposal (which implicitly aligns with Nikhil's take [here](https://github.com/MaterializeInc/materialize/pull/7827#issuecomment-898462738)).

There might be a greatly simplified version that is more like `--drop-recent-ddls 1`, but that kind of destructive operation might lead to some inadvertent heartache, especially given that we can't really differentiate between implicit and explicit DDLs, e.g. `CREATE MATERIALIZED VIEW...` would count as 2: one for the view, one for the index.